python/tchannel/tornado/request.py

```diff
@@ -120,3 +143,38 @@ def get_body_s(self):
         :return: the argstream of body
         """
         return self.argstreams[2]
+
+    def _is_streaming_request(self):
+        """check request is stream request or not"""
+        arg2 = self.argstreams[1]
+        arg3 = self.argstreams[2]
+        return not (isinstance(arg2, InMemStream) and
+                    isinstance(arg3, InMemStream) and
+                    ((arg2.auto_close is True and
+                      arg3.auto_close is True) or (
+                        arg2.state == StreamState.completed and
+                        arg3.state == StreamState.completed)))
+
+    def should_retry_on_error(self, error):
+        """rules for retry"""
+        if self.is_streaming_request:
+            # not retry for streaming request
+            return False
+
+        retry_flag = self.headers.get('re', RetryType.DEFAULT)
+
+        if retry_flag == RetryType.NEVER:
+            return False
+
+        if error.code in [ErrorCode.bad_request, ErrorCode.cancelled]:
```

Comment:
In a different PR, let's change ErrorCode to an enum and include this information in there.

We can define `__init__` methods on `Enum`s to customize behavior. https://docs.python.org/3/library/enum.html#planet